### DISCUSSION TITLE
Add HexBytes field type

### DISCUSSION THREAD

**Discussion by fubuloubu** [pydantic/pydantic#4094] [Feature Request]
It's often very handy to be able to specify fields as being hex-encoded `bytes`, and to be able to serialize them to `bytes` and deserialize them from `bytes`. Pydantic has a lot of handlers for types like filepaths and URLs available, but nothing for hex-encoded bytes.

I propose adding a new type `HexBytes` which works as a thin wrapper around `bytes` with special handling so that it serializes to and deserializes from hex-encoded bytes (with "0x" prefix)

https://github.com/samuelcolvin/pydantic/issues/3756#issuecomment-1129189526 serves as a really solid guide for what this might look like

**Comment by samuelcolvin** [comment]
Same question as base65 - what should the output type be? How should it be encoder by `.dict()` and `.json()`?

This is why we didn't (yet) do a base64 type - everyone wants something else. (I'm on my phone, so can't look for the issue, but it might be worth reading)

**Reply by fubuloubu** [reply]
In both `.dict` and `.json` scenarios, just serialize it back to hex

We were looking into it and it did seem like the best that could be done was to have `.dict` return the class and `.json` have a custom serializer to hex. It might be useful overall to have a way to have custom `.dict` serializer routine for any arbitrary type under `BaseModel._get_value`

**Reply by dmkulazhenko** [reply]
@fubuloubu, or just use bytes as it should be! `HexBytes` is one of the most nasty thing that is used in web3, just useless wrapper that gives nothing but head ache.

**Reply by fubuloubu** [reply]
> @fubuloubu, or just use bytes as it should be! `HexBytes` is one of the most nasty thing that is used in web3, just useless wrapper that gives nothing but head ache. 

We've created `eth-pydantic-types` which adds a `HexBytes` variant that can be used as a pydanticbmodel field and automatically serialized and deserializes to hex string instead of raw bytes

**Reply by dmkulazhenko** [reply]
_Ironic: let's make a thousand useless wrappers, because web3 programming is not at all like web2, ggwp go next :)_

**Reply by fubuloubu** [reply]
It worked for us to get around the issue, so we shared it. This discussion is to try and find a way to resolve it natively within Pydantic so such wrappers are not necessary

**Comment by x11x** [comment]
In case it helps anyone searching for a way to do this -- if you want a simple starting point using Pydantic itself without adding extra dependencies on hexbytes and eth-pydantic-types -- you can use something like the following:
```python
from pydantic import BeforeValidator, PlainSerializer
from annotated_types import Len
from typing import TypeAliasType, Annotated  # (NOTE: use typing_extensions for compatibility)

def parse_hex_bytes(value: str | bytearray | bytes | memoryview) -> bytes:
    # NOTE: we don't do length validation here -- this is done using the Len() annotation
    if isinstance(value, bytes):
        return value
    if isinstance(value, str):
        # NOTE: we don't support padding odd length strings or stripping "0x"-prefix,
        # but this would be simple to add if needed
        return bytes.fromhex(value)
    if isinstance(value, (bytearray, memoryview)):
        return bytes(value)
    # NOTE: we could support int (and its subclass bool) as well if we wanted with
    # if isinstance(value, int):
    #     return value.to_bytes((value.bit_length() + 7) // 8, byteorder='big')
    # This raises OverflowError for negative values, so would need handle this and raise a
    # ValueError to work with Pydantic's validators
    raise TypeError(f"Cannot convert {value!r} of type {type(value).__name__} to bytes")


# Example usage to make a fixed 32 byte type e.g. for a hash:

SHA256_SIZE = 256 // 8


Sha256Sum = TypeAliasType(
    "Sha256Sum",
    Annotated[
        bytes,
        Len(min_length=SHA256_SIZE, max_length=SHA256_SIZE),
        BeforeValidator(parse_hex_bytes),
        PlainSerializer(bytes.hex, return_type=str, when_used='json')  # (could change to when_used='always' if required)
    ]
)
```

NOTE: that this does not left pad too-short strings with zeros like eth-pydantic-types does. This just raises an error if the string is not exactly 64 chars (32 bytes). You could easily left pad the byte strings out to whatever you wanted in the validator (or add another validator that does that) if you wanted to support this.
